keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1783 @@
+[Intel RDT][intel-rdt] implements a class-based mechanism for controlling the
+cache and memory bandwidth QoS of applications. All processes in the same
+hardware class share a portion of cache lines and memory bandwidth. RDT
+provides a way for mitigating noisy neighbors and fulfilling SLAs. In Linux
+control happens via resctrl -- a pseudo-filesystem provided by the kernel which
+makes it virtually agnostic of the hardware architecture. The OCI runtime-spec
+has supported Intel RDT for a while already. Other hardware vendors have
+comparable technologies which use the same [resctrl interface][linux-resctrl].
+
+The Linux Block IO controller parameters depend very heavily on the underlying
+hardware and system configuration (device naming/numbering, IO scheduler
+configuration etc) which makes it very impractical to control from the Pod spec
+level. In order to hide this complexity the concept of blockio classes has been
+added to the container runtimes (CRI-O and containerd). A system administrator
+is able to configure blockio controller parameters on per-class basis and the
+classes are then made available for CRI clients. Following this model also
+provides a possible framework for the future improvements, for instance
+enabling class-based network or memory type prioritization of applications.
+
+Currently, there is no mechanism in Kubernetes to use these types of resources.
+CRI-O and containerd runtimes have support for RDT and blockio classes and they
+provide an bridge-gap user interface through special pod annotations. The goal
+is to get these types of resources first class citizens and properly supported
+in Kubernetes, providing visibility, a well-defined user interface, and
+permission controls.
+
+It seems necessary to support both container-level and pod-level QoS-class
+resources as independent concepts. Intel RDT (above) is per-container by
+design because of the hardware implementation (the control/class hiearchy is
+flat). Also, the current support for blockio is container-level only (it is not
+possible to configure pod sandbox-level cgroup parameters). However, having
+pod-level QoS-class resources makes it possible to implement support for
+sandbox-level blockio parameters. Other usage for pod sandbox-level QoS-class
+resources would be communicating the Kubernetes Pod QoS class from
+kubelet to the container runtime.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Make it possible to request QoS-class resources
+  - Support RDT class assignment of containers. This is already supported by
+    the containerd and CRI-O runtime and part of the OCI runtime-spec
+  - Support blockio class assignment of containers.
+  - Support Pod-level (sandbox-level) QoS-class resources
+- Make the API to support updating QoS-class resource assignment of running containers
+- Make the extensions flexible, enabling simple addition of other QoS-class
```

Comment:
> I think we need to prove that this mechanism is viable for a broader array of resource types, even if we don't fully implement them all.

Mm, agree on this.

> For example, it seems like we should be able to use this to define different levels of memory reclaim aggressiveness

If you refer to `memory.swap.*` controls in cgroups v2 (or other new memory controls in v2) I think this would certainly be a viable usage scenario. This would be "hiding" the actual detailed memory related parameters (configured by the administrator) behind a set of class names. Kubelet would not know anything about the detailed cgroup knobs, just the class name and the container runtime would enforce these via the oci runtime spec (for container cgroups).

> different priority access to "extra" CPUs

Could you elaborate on this a bit? Like priority access to extra cpu cycles or "priority" cpu cores. On top of my head, regarding cpu core priorities, one could e.g. boost exclusively allocated cpus of "high-prio" containers through the systfs cpufreq controls (and throttle "low-prio" ones).
 
> Mostly I just want to make sure we think through how those probably would look - not implement them all.

I think one place where to ponder these could be the *User Stories* section that I commented above (in response to @SergeyKanzhelev) being very much lacking. Need to sleep over this and update the KEP tomorrow...

Reply:
If we could get some plausible-looking manifests and other examples for the different cases, I think that'd help.

Reply:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos sets a throttle factor of 0.8, but maybe "gold" pods get a .8, while "silver" gets 0.5 and "bronze" gets 0.25.

> Kubelet would not know anything about the detailed cgroup knobs, just the class name and the container runtime would enforce these

This is a good point - the aKEP linked above is Kubelet, not the runtime.  It's the "outer" pod cgroup that needs this config.  Should it be possible for kubelet to augment the qos-resources offered by the runtime and add its own?  It would be a shame to add another mechanism to solve the same shape of problem.

Reply:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos sets a throttle factor of 0.8, but maybe "gold" pods get a .8, while "silver" gets 0.5 and "bronze" gets 0.25.

> Kubelet would not know anything about the detailed cgroup knobs, just the class name and the container runtime would enforce these

This is a good point - the aKEP linked above is Kubelet, not the runtime.  It's the "outer" pod cgroup that needs this config.  Should it be possible for kubelet to augment the qos-resources offered by the runtime and add its own?  It would be a shame to add another mechanism to solve the same shape of problem.

Reply:
I think that would be possible and nothing super complex in terms of implementation (when it comes to the scope of this particular KEP i.e. just the API changes). Especially when/if we control/enforce "official" names we could reserve some for kubelet-owned resources.

Actually, this is already envisioned in the KEP under open questions (Pod QoS class) - QoS resources could be used to communicate the Pod QoS class to the runtime. Information they (the runtimes) already want and use but use some kludge-ish cgroup-path-based detection to determine it.

Anyway, I'll mention this as a possible scenario in the KEP. A little bit of side-track (wrt. this KEP), but regarding Pod cgroup handling I've had a secret wish that maybe we could find a way to move that out of kubelet to the runtimes, too 😇